File: bench/closure_error.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func makeAdder(base int) func(int) int {
	return func(x int) int { return base + x }
}

func applyTwice(f func(int) int, x int) int {
	return f(f(x))
}

func benchClosures(iterations int) int {
	total := 0
	for i := 0; i < iterations; i++ {
		add := makeAdder(i)
		step := i%7 + 1
		mul := func(x int) int { return x * step }
		for j := 0; j < 8; j++ {
			total += add(j) - mul(j) + applyTwice(add, j)
		}
	}
	return total
}

func checked(n int) (int, error) {
	label := "value-" + strconv.Itoa(n)
	tags := []string{"a", "b", label}
	if n%5 == 0 {
		return 0, fmt.Errorf("rejected: %s", label)
	}
	return n*2 + len(tags), nil
}

func twoLayers(n int) (int, error) {
	a, err := checked(n)
	if err != nil {
		return 0, err
	}
	b, err := checked(n + 1)
	if err != nil {
		return 0, err
	}
	return a + b, nil
}

func benchErrors(iterations int) int {
	total := 0
	for i := 0; i < iterations; i++ {
		if v, err := twoLayers(i); err == nil {
			total += v
		}
		if v, err := checked(i); err == nil {
			total += v
		} else {
			total += -1
		}
	}
	return total
}

func nowMillis() int64 {
	return time.Now().UnixMilli()
}

func main() {
	iterations := 200000
	if len(os.Args) > 1 {
		if v, err := strconv.Atoi(os.Args[1]); err == nil && v > 0 {
			iterations = v
		}
	}
	fmt.Println("closure/error benchmark")
	fmt.Println("iterations=" + strconv.Itoa(iterations))

	totalStart := nowMillis()

	t0 := nowMillis()
	closureTotal := benchClosures(iterations)
	closureMS := nowMillis() - t0

	t1 := nowMillis()
	errorTotal := benchErrors(iterations)
	errorMS := nowMillis() - t1

	totalMS := nowMillis() - totalStart
	checksum := closureTotal + errorTotal

	fmt.Printf("closure_ms=%d\n", closureMS)
	fmt.Printf("error_ms=%d\n", errorMS)
	fmt.Printf("total_ms=%d\n", totalMS)
	fmt.Printf("checksum=%d\n", checksum)
}

File: bench/closure_error.rs
use std::time::Instant;

fn make_adder(base: i64) -> impl Fn(i64) -> i64 {
    move |x| base + x
}

fn apply_twice<F: Fn(i64) -> i64>(f: &F, x: i64) -> i64 {
    f(f(x))
}

fn bench_closures(iterations: i64) -> i64 {
    let mut total: i64 = 0;
    for i in 0..iterations {
        let add = make_adder(i);
        let step = i % 7 + 1;
        let mul = move |x: i64| x * step;
        for j in 0..8 {
            total += add(j) - mul(j) + apply_twice(&add, j);
        }
    }
    total
}

fn checked(n: i64) -> Result<i64, String> {
    let label = format!("value-{}", n);
    let tags = ["a", "b", label.as_str()];
    if n % 5 == 0 {
        return Err(format!("rejected: {}", label));
    }
    Ok(n * 2 + tags.len() as i64)
}

fn two_layers(n: i64) -> Result<i64, String> {
    let a = checked(n)?;
    let b = checked(n + 1)?;
    Ok(a + b)
}

fn bench_errors(iterations: i64) -> i64 {
    let mut total: i64 = 0;
    for i in 0..iterations {
        total += two_layers(i).unwrap_or(0);
        total += checked(i).unwrap_or(-1);
    }
    total
}

fn main() {
    let iterations: i64 = std::env::args()
        .nth(1)
        .and_then(|a| a.parse().ok())
        .filter(|&v: &i64| v > 0)
        .unwrap_or(200000);
    println!("closure/error benchmark");
    println!("iterations={}", iterations);

    let total_start = Instant::now();

    let t0 = Instant::now();
    let closure_total = bench_closures(iterations);
    let closure_ms = t0.elapsed().as_millis();

    let t1 = Instant::now();
    let error_total = bench_errors(iterations);
    let error_ms = t1.elapsed().as_millis();

    let total_ms = total_start.elapsed().as_millis();
    let checksum = closure_total + error_total;

    println!("closure_ms={}", closure_ms);
    println!("error_ms={}", error_ms);
    println!("total_ms={}", total_ms);
    println!("checksum={}", checksum);
}
